### ISSUE TITLE
Different allocation in memory management between XPU and CUDA

### ISSUE THREAD

**Issue by airMeng** [intel/torch-xpu-ops#1543]
### Discussed in https://github.com/intel/torch-xpu-ops/discussions/1540

<div type='discussions-op-text'>

<sup>Originally posted by **airMeng** April  2, 2025</sup>
In summary, during fine-tuning process, XPU reserved 8GB VRAM more than CUDA, ~~make it impossible to port to BMG-12GB.~~

**This might not be a bug, so I post here for discussion**

Reproduce steps:
1. git clone https://github.com/songhappy/torchtune/tree/debug_memory
2. install the library, prepare the datasets/models as [instructions](https://pytorch.org/torchtune/main/install.html)
3. 
```
tune run lora_dpo_single_device --config recipes/configs/llama3_1/8B_lora_dpo_single_device_my.yaml device=xpu profiler.enabled=False max_steps_per_epoch=1 2>&1
```

What I have tried:
1. torchtune side
I register a hook during backward,                                    
```diff
+        # 🔹 Hook function for gradients
+        def grad_hook(param_name):
+            def hook(grad):
+                torch.xpu.synchronize()
+                peak_memory_reserved = torch_device.max_memory_reserved(self._device) / (1024**3)
+                peak_memory_allocated = torch_device.max_memory_allocated(self._device) / (1024**3)
+                print(f"\n[BACKWARD] Parameter: {param_name} memory reserved", peak_memory_reserved)
+                print(f"\n[BACKWARD] Parameter: {param_name} memory allocated", peak_memory_allocated)
+                return grad
+            return hook
+
+        # 🔹 Register gradient hooks on model parameters (Fix)
+        for name, param in model.named_parameters():
+            if param.requires_grad:
+                param.register_hook(grad_hook(name))
+            else:
+                print(f"Skipping hook for frozen param: {name}")
```     

I also print the memory stats between different steps
```diff
                   loss = loss / self._gradient_accumulation_steps
                   running_loss += loss
+                peak_memory_reserved = torch_device.max_memory_reserved(self._device) / (1024**3)
+                peak_memory_allocated = torch_device.max_memory_allocated(self._device) / (1024**3)
+                print("before backward reserved: ", peak_memory_reserved, " allocated: ", peak_memory_allocated)

                  loss.backward()
+
+                peak_memory_reserved = torch_device.max_memory_reserved(self._device) / (1024**3)
+                peak_memory_allocated = torch_device.max_memory_allocated(self._device) / (1024**3)
+                print("after backward reserved: ", peak_memory_reserved, " allocated: ", peak_memory_allocated)
```

The log shows there are ~8GB allocated even before the first operator got backward
```
before backward reserved:  34.06640625  allocated:  32.82672309875488

[BACKWARD] Parameter: layers.31._checkpoint_wrapped_module.mlp.w2.lora_b.weight memory reserved 41.849609375
```

2. PyTorch Side

I modify the XPU allocator to record each time of the memory allocation:
```diff
diff --git a/c10/xpu/XPUCachingAllocator.cpp b/c10/xpu/XPUCachingAllocator.cpp
index 64eb171616d..fa11702f46c 100644
--- a/c10/xpu/XPUCachingAllocator.cpp
+++ b/c10/xpu/XPUCachingAllocator.cpp
@@ -125,6 +125,8 @@ struct AllocParams {

 } // anonymous namespace

+std::mutex print_mutex;
+
 class DeviceCachingAllocator {
  private:
   mutable std::recursive_mutex mutex;
@@ -265,7 +267,17 @@ class DeviceCachingAllocator {
     p.block = new Block(device, p.queue(), size, p.pool, ptr);
     for_each_selected_stat_type(p.stat_types, [&](size_t stat_type) {
       stats.reserved_bytes[stat_type].increase(size);
-    });
+      std::lock_guard<std::mutex> lock0(print_mutex);
+      printf("%s %lu %s %s %s %s %s %s\n", "size", size, "current ", format_size(stats.reserved_bytes[stat_type].current).c_str(), "peak ", format_size(stats.reserved_bytes[stat_type].peak).c_str(), "allocated ", format_size(stats.reserved_bytes[stat_type].allocated).c_str());
+   });    
   return true;
   }
```

Indeed the extra 8GB allocation is recorded but not reasonable
```
size 4200595456 current  34.02 GiB peak  34.07 GiB allocated  41.99 GiB
size 4200595456 current  33.98 GiB peak  34.03 GiB allocated  41.95 GiB
size 4200595456 current  30.11 GiB peak  34.07 GiB allocated  45.90 GiB
size 4200595456 current  30.07 GiB peak  34.03 GiB allocated  45.86 GiB
size 4200595456 current  34.02 GiB peak  34.07 GiB allocated  49.81 GiB
size 4200595456 current  33.98 GiB peak  34.03 GiB allocated  49.77 GiB
size 4202692608 current  37.94 GiB peak  37.94 GiB allocated  53.72 GiB
size 4202692608 current  37.90 GiB peak  37.90 GiB allocated  53.69 GiB
size 4202692608 current  41.85 GiB peak  41.85 GiB allocated  57.64 GiB
size 4202692608 current  41.81 GiB peak  41.81 GiB allocated  57.60 GiB
size 2097152 current  41.85 GiB peak  41.85 GiB allocated  57.64 GiB
size 2097152 current  42.00 MiB peak  42.00 MiB allocated  42.00 MiB
size 2097152 current  41.85 GiB peak  41.85 GiB allocated  57.64 GiB
size 2097152 current  44.00 MiB peak  44.00 MiB allocated  44.00 MiB
size 2097152 current  41.86 GiB peak  41.86 GiB allocated  57.64 GiB
size 2097152 current  46.00 MiB peak  46.00 MiB allocated  46.00 MiB
```

Note that 4202692608 is 4008\*1024\*\*2, 4200595456 is 4006\*1024\*\*2, and there are no tensors that has shapes divided or multiplied by 4006 or 4008.

3. CUDA side
You can see no extra memory reserved before the the first layer got backward.
```
 390 before backward reserved 36.376953125
 391 before backward allocated 32.83465766906738
 392
 393 [BACKWARD] Parameter: layers.31._checkpoint_wrapped_module.mlp.w2.lora_b.weight reserved 36.376953125
 394
 395 [BACKWARD] Parameter: layers.31._checkpoint_wrapped_module.mlp.w2.lora_b.weight allocated 32.83465766906738
```</div>

**Comment by chuanqi129** [comment]
@guangyey any comment for this issue?

**Comment by guangyey** [comment]
There are some subtle differences between the XPU and CUDA allocators. We are currently working on refactoring the XPU allocator to align with the same core logic, with the expectation of completing this before Q3.

**Comment by guangyey** [comment]
If you could provide a mini reproducer, it would facilitate our hot fix on it.

**Comment by ppapierk** [comment]
Hi @airMeng @riverliuintel,

I need support with reproduction of this issue to fix it. Please help.

**Comment by ppapierk** [comment]
hi @airMeng 
Thanks for response. What about 8B_lora_dpo_single_device**_my**.yaml? Does it has any other things modified except from target device type?

**Comment by airMeng** [comment]
> hi [@airMeng](https://github.com/airMeng) Thanks for response. What about 8B_lora_dpo_single_device**_my**.yaml? Does it has any other things modified except from target device type?

no more changes

**Comment by ppapierk** [comment]
Memory allocations described in bug are comming from backward propagation of  `logits_to_logprobs' function in `torchtune/rlhf/sequence_processing.py` file.

```

def logits_to_logprobs(
    logits: torch.Tensor, sequences: torch.Tensor, temperature: float = 1.0
) -> torch.Tensor:
    """
    Converts logits corresponding to a generated sequence to logprobs over the generated tokens.

    Args:
        logits (torch.Tensor): The logits tensor of shape [b, response_length, vocab_size].
        sequences (torch.Tensor): The corresponding tokens of shape [b, response_length].
        temperature (float): The temperature to scale the logits. Default 1.0
    Returns:
        torch.Tensor: The log probabilities corresponding to each token in ``sequences``. Shape [b, response_length].
    """

    x = torch.gather(
        F.log_softmax(logits / temperature, dim=-1),
        2,
        sequences.unsqueeze(-1),
    ).squeeze(-1)

    return x
```

These allocation are for intermediate tensors and are released immediately after. To by more precise, in forward:
- output for div (logits / temperature) operator is allocated and div is executerd
- output for log_softmax is allocated, log_softmax is executed and output for div is freed
- output for gather is allocated, gather is executed and output for log_softmax is freed

Similar flow happens at the beginning of the backward pass, this is were highest peak reserved memory consumption is observed. This memory however does not reduce reserved device memory pool, because of how Caching Allocator works. Buffers  (memory blocks) are kept allocated for further usage and freed only when user specifically request that or we run out of free memory. 

So for backward pass most of the already allocated memory blocks should be reused. In tested case however on backward we do not have proper memory block ready and ...
--- to be continued --
